Make NetCDF file cache handling compatible with dask distributed

gerritholl opened the pull request:
This PR makes file cache handling in the NetCDF4FileHandler compatible with dask distributed.  It adds a utility function in `satpy.readers.utils` called `get_distributed_friendly_dask_array`, which can be used to produce a dask.array from a netCDF4 variable that can be used in an xarray, but dask graphs remain picklable and thus computable when including this one.  This utility function is now used in NetCDF4FileHandler, which replaces homegrown file handle caching by caching using xarray.backends.CachingFileManager, which is needed to implement the aforementioned utility function.

 - [x] Closes #2815 <!-- remove if there is no corresponding issue, which should only be the case for minor changes -->
 - [x] Tests added <!-- for all bug fixes or enhancements -->
 - [x] Fully documented <!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->

djhoese committed:
Empty commit for testing

djhoese committed:
Trigger PR checks after CodeFactor bug report closed

djhoese committed:
Trigger PR checks after CodeFactor bug report closed

djhoese committed:
Retrigger codefactor analysis

djhoese committed:
Retrigger tests to prove caching works

strandgren committed:
Add tests for default functionality of HighCloudCompositor and LowCloudCompositor.

gerritholl committed:
Add test to reproduce GH 2815

gerritholl committed:
make sure distributed client is local

gerritholl committed:
Start utility function for distributed friendly

Start work on a utility function to get a dask array from a dataset
variable in a way that is friendly to dask.distributed.

gerritholl committed:
Parameterise test and simplify implementation

For the distributed-friendly dask array helper, parameterise the test
to cover more cases.  Simplify the implementation.

gerritholl committed:
Force shape and dtype.  First working prototype.

We need to force the shape and the dtype when getting the
dask-distributed-friendly xarray-dataarray.  Seems to have a first
working prototype now.

gerritholl committed:
Add group support and speed up tests

Add group support for getting a dask distributed friendly dask array.
Speed up the related tests by sharing the dask distributed client setup
and breakdown.

gerritholl committed:
Add partial backward-compatibility fol file handle

Add partial backward compatibility for accessing the file handle
attribute when using caching with a NetCDF4FileHandler base class.
Backward incompatibility is not 100%.  Deleting the FileHandler closes
the manager and therefore the ``file_handle`` property, however, when
accessing the ``file_handle`` property after deleting the
``FileHandler``, it is reopened.  Therefore, calling `__del__()``
manually and then accessing ``fh.file_handle`` will now return an open file
(was a closed file).  This should not happen in any sane use scenario.

gerritholl committed:
Respect auto_maskandscale with new caching

With the new dask-distributed-friendly caching, make sure we are
respecting auto_maskandscale and are not applying scale factors twice.

gerritholl committed:
Remove needless except block

Remove a dead code except block that should never be reached.

gerritholl committed:
Test refactoring

Migrate TestNetCDF4FileHandler from unittest.TestCase to a regular
class.  Use a pytest fixture for the temporary NetCDF file.

djhoese commented on satpy/readers/netcdf_utils.py:
```diff
@@ -196,11 +210,8 @@ def _get_required_variable_names(listed_variables, variable_name_replacements):
-            except RuntimeError:  # presumably closed already
-                pass
+        if self.manager is not None:
+            self.manager.close()
```
Is it worth it to wrap this with a try/except?

gerritholl replied:
Maybe?  I figured the `RuntimeError` could not happen because the `CachingFileManager` does not fail when close is called multiple times.  That leaves other exceptions, but I'm not sure what we would do in the `except:`-block, nor what actually happens if an exception is uncaught in `__del__`.

djhoese replied:
Yeah I'm not sure either. Let's see what happens I guess.

djhoese commented on satpy/readers/utils.py:
```diff
@@ -474,3 +475,53 @@ def remove_earthsun_distance_correction(reflectance, utc_date=None):
+
+    Example::
+
+        >>> import NetCDF4
+        >>> from xarray.backends import CachingFileManager
+        >>> cfm = CachingFileManager(NetCDF4.Dataset, fn, mode="r")
+        >>> arr = get_distributed_friendly_dask_array(cfm, "my_var")
```
```suggestion
        >>> import netCDF4
        >>> from xarray.backends import CachingFileManager
        >>> cfm = CachingFileManager(netCDF4.Dataset, fn, mode="r")
        >>> arr = get_distributed_friendly_dask_array(cfm, "my_var")
```

djhoese commented on satpy/readers/utils.py:
```diff
@@ -474,3 +475,53 @@ def remove_earthsun_distance_correction(reflectance, utc_date=None):
+        >>> arr = get_distributed_friendly_dask_array(cfm, "my_var")
+
+    Args:
+        manager (xarray.backends.CachingFileManager):
+            Instance of xarray.backends.CachingFileManager encapsulating the
+            dataset to be read.
```
We should check how the docs render this. If the argument type isn't "clickable" to go directly to the xarray docs for the CFM then we could wrap the mention of it in the description with:

```
:class:`xarray.backends.CachingFileManager`
```

gerritholl replied:
The argument type was already clickable, but in the description it was not.  I have now made it clickable in both cases (screenshot from local doc production):
 
![Bildschirmfoto_2024-07-25_11-10-06](https://github.com/user-attachments/assets/bcc59e65-6ddc-4716-9296-e729f3b8ff1b)

djhoese commented on satpy/readers/utils.py:
```diff
@@ -474,3 +475,53 @@ def remove_earthsun_distance_correction(reflectance, utc_date=None):
     return reflectance
+
+
+def get_distributed_friendly_dask_array(manager, varname, chunks, dtype,
+                                        group="/", auto_maskandscale=None):
```
I'm not sure how I feel about this function name. Obviously it makes sense in this PR because it solves this specific problem, but it feels like there is a (shorter) more generic name that gets the point across. Another thing is that `distributed_friendly` is mentioned here, but that friendliness is a side effect of the "serializable" nature of the way you're accessing the data here, right? `get_serializable_dask_array`?

I don't feel super strongly about this, but the name was distracting to me so I thought I'd say something.

gerritholl replied:
Renamed `get_serializable_dask_array`.

djhoese commented on satpy/readers/utils.py:
```diff
@@ -474,3 +475,53 @@ def remove_earthsun_distance_correction(reflectance, utc_date=None):
+    """
+    def get_chunk():
+        with manager.acquire_context() as nc:
+            if auto_maskandscale is not None:
+                nc.set_auto_maskandscale(auto_maskandscale)
+            return nc["/".join([group, varname])][:]
```
This code is specific to `netCDF4`. While I understand that is the use case here, it suggests that either the function name and purpose needs to change or the `set_auto_maskandscale` functionality should be changed. Along with that, if I'm understanding the CFM properly, the `nc` object here should be the same between accesses. I mean, the exact same object if accessed in the same process. Although unlikely in the current usage setting this global attribute on the `netCDF4.Dataset` object could cause race conditions or unexpected behavior...although very unlikely.

To resolve both of these concerns, could this function and the current usage be changed to:

1. Use a wrapper (or `functools.partial` is possible) around `netCDF4.Dataset` when creating the CFM to set the auto_maskandscale option when `netCDF4.Dataset` is called inside the CFM. Although more hacky, there could also be a callback/init_func or maybe getitem function (see next 2 points) that handles setting of these per-variable things there (auto mask and scale can be set on the file object or the variable object).
2. The `group` and `varname` being joined is NetCDF specific and as far as I can tell isn't needed if the user is just expected to do it themselves.
3. This function assumes that `__getitem__` works for whatever file-like object is accessed and that `[:]` is needed. I don't think `__getitem__/[]` access for a variable is a bad assumption, but if there is a way to make this generic maybe that's what we need. If this function ended up being used with another library for reading files then the `[:]` may make an unnecessary copy and may not be noticeable to someone using the function.

gerritholl replied:
I've added a wrapper along with `functools.partial` and moved the `set_auto_maskandscale` there.  I've only set this globally; I find per variable too complicated and we set it globally so far already, so there is no change there.  I've also changed the group handling and moved this responsibility to the user.

As for the `__getitem__` assumption, yes, there is such an assumption.  I'm not sure how to make this generic without making it harder to use.  To make it generic, the user would need to pass information in some way on how the variable is obtained, via a callable or a wrapper class.  Do you have an idea on what would be a good way to implement this?

djhoese replied:
I'm ok with the `__getitem__` assumption and even the `[:]` if it just seems annoying to make it more generic. One option could be to do something like `if not isinstance(x, np.ndarray)` (or some other "is not numpy-like or is netcdf Variable-like object" type of check) then you could do the `[:]` inside that `if` block.

gerritholl replied:
I replaced `[:]` by taking information from `block_info`:

```
    def get_chunk(block_info=None):
        arrloc = block_info[None]["array-location"]
        with manager.acquire_context() as nc:
            var = nc[varname]
            return var[tuple(slice(*x) for x in arrloc)]
```

It's less obvious than `[:]` (at least to me) but I think it's correct :D

If something doesn't support such indexing, then it will fail with an exception and the user will notice that it doesn't work.

djhoese replied:
Sounds good.

djhoese commented on satpy/readers/utils.py:
```diff
@@ -474,3 +475,53 @@ def remove_earthsun_distance_correction(reflectance, utc_date=None):
+            method set_auto_maskandscale, such as is the case for
+            NetCDF4.Dataset.
+    """
+    def get_chunk():
```
The `chunks` is never used here. The current calling from the file handler is accessing the full shape of the variable so this is fine, but only for now. I mean that `map_blocks` will only ever call this function once. *However*, if you added a `block_info` kwarg to the function signature or whatever the `map_blocks` special keyword argument is, then you could change `[:]` to access a specific sub-set of the NetCDF file variable and only do a partial load. This should improve performance a lot (I think :crossed_fingers:) if it was actually used in the file handler.

djhoese replied:
Yes I think that's what I'm saying. I think the result of `get_chunk()` right now is broken for any chunk size other than the full shape of the array because you never do any slicing of the NetCDF variable inside `get_chunk()`. So, if you had a full array of 100x100 and a chunk size of 50x50, then map_blocks would call this function 4 times ((0-50, 0-50), (0-50, 50-100), (50-100, 0-50), (50-100, 50-100)). BUT each call would return the full variable 100x100. So I *think* this would be a case where the dask array would say "yeah, I have shape 100x100", but then once you computed it you'd get a 200x200 array back.

gerritholl replied:
Fixed it now, I think.

djhoese commented on satpy/readers/utils.py:
```diff
@@ -474,3 +475,53 @@ def remove_earthsun_distance_correction(reflectance, utc_date=None):
+            get_chunk,
+            chunks=chunks,
+            dtype=dtype,
+            meta=np.array([]))
```
I think `dtype=dtype` should be added to the `np.array` call too. Otherwise it looks like numpy defaults to `[]` being a `np.float64` array.

djhoese commented on satpy/readers/netcdf_utils.py:
```diff
@@ -85,10 +85,12 @@ class NetCDF4FileHandler(BaseFileHandler):
         xarray_kwargs (dict): Addition arguments to `xarray.open_dataset`
         cache_var_size (int): Cache variables smaller than this size.
         cache_handle (bool): Keep files open for lifetime of filehandler.
+            Uses xarray.backends.CachingFileManager, which uses a least
+            recently used cache.
```
I've lost track of the caching functionality of this helper class. What are the main differences now between the xarray implementation and the "cached" version specific to Satpy? Is the main difference the support for NetCDF groups and loading per-variable?

gerritholl replied:
`CachingFileManager` from xarray uses (by default) xarrays global LRU file cache, which by default keeps at most 128 files open (configurable).  Satpy kept everything open indefinitely. This could have an impact on performance, but is probably a good thing overall. 

By using `CachingFileManager`, the resulting `xarray.DataArray` is serialisable and therefore dask-distributed friendly.

Not sure about your point on groups — groups were already supported?

djhoese replied:
I didn't think xarray's default `xr.open_dataset` supported groups so we had this helper class and added on caching functionality later. I think the idea I was going for when I made the above comment originally was: why can't we just use `xr.open_dataset` instead of this class using `CachingFileManager`? My assumption was that we're supporting NetCDF groups here while xarray doesn't.

gerritholl replied:
Oh, do you mean the entire `NetCDF4FileHandler` class is redundant if we use `CachingFileManager` and `xr.open_dataset`s group ability?  It would seem to be a bit more, such as providing a convenient indexing mechanism.  But maybe I still misunderstand.

djhoese replied:
I guess that's what I'm asking, not stating. You're right there is the `__getitem__` logic and the group handling is important enough. I guess I was mostly just curious what the difference was overall.

djhoese reviewed (changes requested):
Awesome job deciphering how to use the CachingFileManager and wrapping things in a map_blocks task. I think this is really close to being done, but I had some concerns about the helper function.

gerritholl committed:
fix docstring example spelling

Fix the spelling in the docstring example using netCDF4.

Co-authored-by: David Hoese <david.hoese@ssec.wisc.edu>

gerritholl committed:
Prevent unexpected type promotion in unit test

Add a workaround to prevent an unexpected type promotion in the unit
test for dask distributed friendly dask arrays.

gerritholl committed:
Use block info getting a dd-friendly da

When getting a dask-distributed friendly dask array from a NetCDF file
using the CachingFileManager, use the information provided in bloc_info
on the array location in case we are reading not the entire variable.

gerritholl committed:
Rename to serialisable and remove group argument

Rename get_distributed_friendly_dask_array to
get_serialisable_dask_array and remove the group argument, moving the
responsibility for handlings groups to the caller.

gerritholl committed:
Use wrapper class for auto_maskandscale

gerritholl committed:
GB -> US spelling

Pytroll uses US spelling.  Rename serializable to serialisable.

Remove removed keyword argument from call.

gerritholl committed:
Ensure meta dtype

Ensure that the meta we pass to map_blocks also has the right dtype.
Not sure if this is necessary when map_blocks already has the right
dtype, but it can't hurt.

gerritholl committed:
Fix spelling in test

mraspaud commented on satpy/readers/utils.py:
```diff
@@ -476,6 +477,48 @@ def remove_earthsun_distance_correction(reflectance, utc_date=None):
+def get_serializable_dask_array(manager, varname, chunks, dtype):
+    """Construct a serializable dask array from a variable.
+
+    When we construct a dask array using da.array and use that to create an
```
Should we specify that this happens when a dask array is constructed from a file?

gerritholl replied:
Yes.  Fixed in 5b137e8e7.

mraspaud reviewed (approved):
LGTM, just one comment inline. How is this affecting the performance of the reader? Do you consider this ready to be merged?

gerritholl committed:
Clarify docstring

gerritholl commented:
> How is this affecting the performance of the reader?

I had not tested that yet.  I have now.  Sadly, it gets worse :(

A simple test script that reads FCI, loads some channels, resamples, and writes them again, not specifying any dask scheduler.

With Satpy main:

time 0:45.93, RAM 9.27 GB

With this PR:

time 0:55.0, RAM 10.0 GB

Additionally, upon exiting, there is the repeated error message:

```
 Original exception was:
Error in sys.excepthook:

Original exception was:
Error in sys.excepthook:

Original exception was:
Error in sys.excepthook:
```

> Do you consider this ready to be merged?

Sadly no, considering the problems described above.  I will dig into this.

gerritholl commented:
With Satpy main, three runs, times in seconds:

Scene creation: 13.9, 11.9, 11.5

Loading: 1.5, 1.3, 1.0

Computing: 6.5, 6.9, 6.5

With this PR:

Scene creation: 13.2,  12.0, 12.0

Loading: 3.5, 3.5, 4.4

Computing: 5.5, 5.8, 5.6

So it's in particular the loading that gets slower.

gerritholl commented:
Profiling reveals that there are 160 calls to `acquire_context`, but only 40 to `_NCDatasetWrapper.__init__`, so the caching appears to be doing its job.  But the load call is not reusing the opening that happened upon Scene creation, which is why performance is impacted.  I will see what I can improve.

gerritholl committed:
Use cache already in scene creation

When caching, make sure we use the CachingFileManager already upon scene
creation and not only by the time we are loading.

gerritholl commented:
With c2b1533 loading is much faster, although with more variability than for satpy main.  Scene creation is a little slower.

Scene creation: 13.4, 14.5, 12.8, 12.6, 12.8, 13.0

Loading: 1.9, 1.0, 1.6, 1.4, 1.3, 1.0

gerritholl commented:
I can't reliably reproduce the performance differences.  Running it again with the main branch gives:

Scene creation, main branch: 14.5, 14.3, 13.2

Scene creation, this PR: 13.6, 12.7, 13.8

And with cProfile, it's always faster with my PR.

Considering those uncertainties, I will declare performance is the same within the measurement uncertainty.

gerritholl committed:
Use helper function rather than subclass

Don't subclass netCDF4.Dataset, rather just return an instance from a
helper function.  Seems good enough and gets rid of the weird error
messages upon exit.

gerritholl commented:
Fixed the problem with the strange exception/error messages upon exit in 9fce5a7.

gerritholl committed:
restore non-cached group retrieval

Some readers read entire groups; this needs xarray kwargs to be set even
if caching is used.

mraspaud commented:
I'm happy with this. @djhoese can you just confirm that you are good with this being merged? (and feel free to merge it if that's the case)

djhoese commented on satpy/readers/netcdf_utils.py:
```diff
@@ -117,14 +128,22 @@ def __init__(self, filename, filename_info, filetype_info,
+
     @staticmethod
     def _set_file_handle_auto_maskandscale(file_handle, auto_maskandscale):
         if hasattr(file_handle, "set_auto_maskandscale"):
```
Not that this has to be handled in your PR, but if I remember correctly this `Dataset`-level `set_auto_maskandscale` was added to netcdf4-python quite a while ago. It seems error prone and confusing to silently call the method only if it exists and to not log/inform the user that it wasn't used when it was expected. Maybe we should remove this method on the file handler class and always call `file_handle.set_auto_maskandscale` no matter what. Your wrapper does it already.

gerritholl replied:
If it doesn't need to be handled in this PR then I'd rather defer it.

djhoese commented on satpy/tests/reader_tests/test_utils.py:
```diff
@@ -514,6 +514,64 @@ def test_generic_open_binary(tmp_path, data, filename, mode):
+    slow.
+    """
+
+    @pytest.fixture(scope="class")
+    def dask_dist_client(self):
+        """Set up and close a dask distributed client."""
+        from dask.distributed import Client
+        cl = Client()
+        yield cl
+        cl.close()
```
It looks like dask developer's recommend using their test utilities for writing distributed-based tests:

https://distributed.dask.org/en/latest/develop.html#writing-tests

Would it be possible to use their tools instead of this?

djhoese replied:
Looks like you might be able to import this fixture and use it:

https://github.com/dask/distributed/blob/386e5fea1cde4aefaf821e319405188266b41832/distributed/utils_test.py#L521-L525

gerritholl replied:
Fixed in 222648afe

gerritholl committed:
use fixture not fixture()

Fix an error reintroduced when rebasing and solving merge conflicts.
Use fixture not fixture().

gerritholl committed:
Use dask distributed fixtures

gerritholl commented:
`ruff check --fix` and `isort` are reverting each others changes to the list of imports.  I don't know how to satisfy both.

gerritholl committed:
satisfy ruff, desatisfy isort

gerritholl committed:
avoid renaming on import

gerritholl commented:
I think CodeScene is not fair in complaining about the excess number of arguments:

```def test_get_serializable_dask_array(self, tmp_path, client, shape, dtype, grp):```

Two of those are fixtures, so there is no way I can get rid of them.  I don't think combining shape, dtype, grp is going to improve this method in any way.  I tried to suppress the finding, but it doesn't work:

> Error occurred while suppressing! Please try again.

gerritholl committed:
Remove class for get serializable dask array test

Remove the class around the get_serializable_dask_array test, that is no
longer doing anything now that we get the fixture from dask.

gerritholl commented:
> You could remove the surrounding class as it isn't doing anything. That would get rid of the `self` argument too.

Good point, done.  But CodeScene already discounts `self`, so it doesn't resolve the failure.

gerritholl commented on satpy/readers/core/netcdf.py:
```diff
@@ -196,11 +215,8 @@ def _get_required_variable_names(listed_variables, variable_name_replacements):
-            except RuntimeError:  # presumably closed already
-                pass
+        if self.manager is not None:
+            self.manager.close()
```
I don't think so, but this is `__del__` — exceptions are swallowed anyway?  And what good would it do to change the value of `self.manager` if the object is under destruction?

djhoese replied:
I think I was thinking of race conditions in a multi-threaded environment. I know we've seen users get errors where `__del__` is called on a file handler that is already closed or deleted and they get NetCDF errors or something. Anyway...I guess leave it as is and we'll see what happens.

djhoese commented on satpy/readers/core/utils.py:
```diff
@@ -699,3 +700,45 @@ def get_coefs(self, channel):
+        dtype (dtype):
+            What dtype to use.
+    """
+    def get_chunk(block_info=None):
+        arrloc = block_info[None]["array-location"]
+        with manager.acquire_context() as nc:
+            var = nc[varname]
+            return var[tuple(slice(*x) for x in arrloc)]
```
I'm not sure where our discussion ended on this, but I still think `manager` and `varname` could/should be keyword arguments to this function. Oh actually, for full distributed compatibility I think this `get_chunk` function needs to be in the global/module namespace otherwise distributed can't serialize it easily or at all.

mraspaud commented:
Should we suggest this change to xarray, or is this too specific?
